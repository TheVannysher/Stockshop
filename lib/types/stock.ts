export type SocialMediaCount = {
  type: string,
  count: number,
}

export type Stock = {
  symbol: string;
  // average buying price from buying positions in the order book
  buyPrice: number;
  // average selling price from selling positions in the order book
  sellPrice: number;
  socialMedia?: SocialMediaCount[]
  date: Date;
}

export interface Order {
  type: 'sell' | 'buy',
  stock: Stock,
  quantity: number,
}
