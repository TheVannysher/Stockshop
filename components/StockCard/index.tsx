import React, { useMemo } from 'react'
import { StyleSheet, View } from 'react-native'
import { Text } from '../Themed';
import Badge from '../badge';
import Card from '../card';
import { SocialMediaCount } from '@/lib/types/stock';

export interface StockCardProps {
  symbol: string;
  sellPrice: number;
  buyPrice: number;
  recommandation: string;
  date: Date;
  socials?: SocialMediaCount[],
}

export default function StockCard(props: StockCardProps) {
  const {
    symbol,
    sellPrice,
    buyPrice,
    recommandation,
    date,
    socials,
  } = props;

  const price = useMemo(() => {
    return `${parseFloat(((buyPrice + sellPrice) / 2).toString()).toFixed(2)}$`;
  }, [buyPrice, sellPrice]);

  return (
    <Card
      title={symbol}
      badge={
        <View style={styles.entry}>
          <Text style={styles.socialTitle}>current price: </Text>
          <Badge title={price} />
        </View>
      }
    >
      <View style={{ marginBottom: 10 }}>
        <View style={styles.entry}>
          <Text>ask:</Text>
          <Text>{parseFloat(buyPrice.toString()).toFixed(2)}$</Text>
        </View>
        <View style={styles.entry}>
          <Text>bid:</Text>
          <Text>{parseFloat(sellPrice.toString()).toFixed(2)}$</Text>
        </View>
        {socials && (
          <Text style={styles.socialTitle}>Mentions:</Text>
        )}
        {socials && socials.map((social) => (
          <View key={social.type} style={styles.entry}>
            <Text>{social.type}</Text>
            <Text>{parseInt(social.count.toString())}</Text>
          </View>
        ))}
      </View>
      <View style={{
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
        gap: 5,
      }}>
        <View style={styles.buttonContainer}>
          <Text>Recommandation:</Text>
          {recommandation === 'buy' && <Badge title="buy" color="#3EC300" />}
          {recommandation === 'sell' && <Badge title="sell" color="#E13700" />}
          {recommandation === 'hold' && <Badge title="hold" />}
        </View>
        <View style={styles.entry}>
          <Text>on: </Text>
          <Text>{date.toLocaleDateString()}</Text>
        </View>
      </View>
    </Card>
  )
}

const styles = StyleSheet.create({
  buttonContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 5,
    padding: 5,
  },
  container: {
    flex: 1,
    padding: 5,
    gap: 10,
  },
  entry: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  socialTitle: { padding: 5, fontWeight: 'bold' },
});