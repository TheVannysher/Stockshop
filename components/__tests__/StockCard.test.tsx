import * as React from 'react';
import renderer from 'react-test-renderer';

import StockCard from '../StockCard';
import { Stock } from '@/lib/types/stock';

it(`renders correctly`, () => {
  const stock: Stock = {
    symbol: 'AAPL',
    buyPrice: 100,
    sellPrice: 100,
    date: new Date(),
    socialMedia: [
      {
        type: 'twitter',
        count: 540,
      }
    ]
  }
  const tree = renderer.create(
    <StockCard
      symbol={stock.symbol}
      buyPrice={stock.buyPrice}
      date={stock.date}
      sellPrice={stock.sellPrice}
      recommandation={'buy'}
      socials={stock.socialMedia}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`renders correctly`, () => {
  const stock: Stock = {
    symbol: 'WWWWWWWWWWWWWWWWWW',
    buyPrice: 10000,
    sellPrice: 1000000,
    date: new Date(),
    socialMedia: [
      {
        type: 'WWWWWWWWW',
        count: 54000000,
      }
    ]
  }
  const tree = renderer.create(
    <StockCard
      symbol={stock.symbol}
      buyPrice={stock.buyPrice}
      date={stock.date}
      sellPrice={stock.sellPrice}
      recommandation={'buy'}
      socials={stock.socialMedia}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
