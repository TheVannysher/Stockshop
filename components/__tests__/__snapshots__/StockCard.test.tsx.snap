// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
        "borderRadius": 8,
        "borderTopColor": "#2f95dc",
        "borderTopWidth": 4,
        "padding": 10,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 10,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 10,
      },
      {
        "borderTopColor": "#2f95dc",
      },
    ]
  }
>
  <View>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#000",
            },
            {
              "fontSize": 20,
              "fontWeight": "bold",
            },
          ]
        }
      >
        AAPL
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "gap": 2,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 100,
                "justifyContent": "center",
                "padding": 5,
              },
              {
                "backgroundColor": "#2f95dc",
              },
            ]
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontWeight": "bold",
                "textTransform": "uppercase",
              }
            }
          >
            100.00$
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        {
          "marginBottom": 10,
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          ask:
        </Text>
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          100.00
          $
        </Text>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          bid:
        </Text>
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          100.00
          $
        </Text>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          date:
        </Text>
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          3/11/2024
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "#000",
            },
            {
              "fontWeight": "bold",
              "padding": 5,
            },
          ]
        }
      >
        Mentions:
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          twitter
        </Text>
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          540
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "gap": 5,
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 5,
            "padding": 5,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          Recommandation:
        </Text>
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 100,
                "justifyContent": "center",
                "padding": 5,
              },
              {
                "backgroundColor": "#3EC300",
              },
            ]
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontWeight": "bold",
                "textTransform": "uppercase",
              }
            }
          >
            buy
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly 2`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
        "borderRadius": 8,
        "borderTopColor": "#2f95dc",
        "borderTopWidth": 4,
        "padding": 10,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 10,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 10,
      },
      {
        "borderTopColor": "#2f95dc",
      },
    ]
  }
>
  <View>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#000",
            },
            {
              "fontSize": 20,
              "fontWeight": "bold",
            },
          ]
        }
      >
        WWWWWWWWWWWWWWWWWW
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "gap": 2,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 100,
                "justifyContent": "center",
                "padding": 5,
              },
              {
                "backgroundColor": "#2f95dc",
              },
            ]
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontWeight": "bold",
                "textTransform": "uppercase",
              }
            }
          >
            505000.00$
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        {
          "marginBottom": 10,
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          ask:
        </Text>
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          10000.00
          $
        </Text>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          bid:
        </Text>
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          1000000.00
          $
        </Text>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          date:
        </Text>
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          3/11/2024
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "#000",
            },
            {
              "fontWeight": "bold",
              "padding": 5,
            },
          ]
        }
      >
        Mentions:
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          WWWWWWWWW
        </Text>
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          54000000
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "gap": 5,
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 5,
            "padding": 5,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          Recommandation:
        </Text>
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 100,
                "justifyContent": "center",
                "padding": 5,
              },
              {
                "backgroundColor": "#3EC300",
              },
            ]
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontWeight": "bold",
                "textTransform": "uppercase",
              }
            }
          >
            buy
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
